packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts

```diff
@@ -0,0 +1,139 @@
+import {EntryPoint, INVALID_ENTRY_POINT, NO_ENTRY_POINT, getEntryPointInfo} from './entry_point';
+
+/**
+ * Manages reading a write a manifest file that contains a list of all the entry-points that were
+ * found below a give basePath.
+ *
+ * This is as super-set of the entry-points that are actually processed for a given run of ngcc,
+ * since some may already be processed, or excluded if they do not have the required format.
+ */
+export class EntryPointManifest {
+  constructor(private fs: FileSystem, private config: NgccConfiguration, private logger: Logger) {}
+
+  /**
+   * Try to get the entry-point info from a manifest file for the given `basePath` if it exists and
+   * is not out of date.
+   *
+   * Reasons for the manifest to be out of date are:
+   *
+   * * the file does not exist
+   * * the ngcc version has changed
+   * * the package lock-file (i.e. yarn.lock or package-lock.json) has changed
+   * * the project configuration has changed
+   * * one or more entry-points in the manifest are not valid
+   *
+   * @param basePath The path that would contain the entry-points and the manifest file.
+   * @returns an array of entry-point information for all entry-points found below the given
+   * `basePath` or `null` if the manifest was out of date.
+   */
+  readEntryPointsUsingManifest(basePath: AbsoluteFsPath): EntryPoint[]|null {
+    try {
+      if (this.fs.basename(basePath) !== 'node_modules') {
+        return null;
+      }
+
+      const computedLockFileHash = this.computeLockFileHash(basePath);
+      if (computedLockFileHash === null) {
+        return null;
+      }
+
+      const manifestPath = this.getEntryPointManifestPath(basePath);
+      if (!this.fs.exists(manifestPath)) {
+        return null;
+      }
+
+      const {ngccVersion, configFileHash, lockFileHash, entryPointPaths} =
+          JSON.parse(this.fs.readFile(manifestPath)) as EntryPointManifestFile;
+      if (ngccVersion !== NGCC_VERSION || configFileHash !== this.config.hash ||
+          lockFileHash !== computedLockFileHash) {
+        return null;
+      }
+
+      this.logger.debug(
+          `Entry-point manifest found for ${basePath} so loading entry-point information directly.`);
+      const startTime = Date.now();
+
+      const entryPoints: EntryPoint[] = [];
+      for (const [packagePath, entryPointPath] of entryPointPaths) {
+        const result =
+            getEntryPointInfo(this.fs, this.config, this.logger, packagePath, entryPointPath);
+        if (result === NO_ENTRY_POINT || result === INVALID_ENTRY_POINT) {
+          throw new Error(
+              `The entry-point manifest at ${manifestPath} contained an invalid pair of package paths: [${packagePath}, ${entryPointPath}]`);
+        } else {
+          entryPoints.push(result);
+        }
+      }
+      const duration = Math.round((Date.now() - startTime) / 100) / 10;
+      this.logger.debug(`Reading entry-points directly took ${duration}s.`);
```

Comment:
nit: Mentioning _manifest_ here is more clear than _directly_

```suggestion
      this.logger.debug(`Reading entry-points from manifest took ${duration}s.`);
```

Reply:
Strictly the entry-points were not read from the manifest, only their paths are read from the manifest. We still read the entry-point info from the files on disk.

Reply:
```suggestion
      this.logger.debug(`Reading entry-points using the manifest entries took ${duration}s.`);
```

Reply:
Oh, yeah that is true. I was thinking that if I were to read that debug message, I wouldn't know what _directly_ would actually mean.